File: java/1345.跳跃游戏.java
class Solution {
    // 1. O(n) 83ms(5%) O(n) 52.9M(57%)
    public int minJumps(int[] arr) {
        Map<Integer, List<Integer>> idxMap = new HashMap<>();
        int n = arr.length;
        int step = 0;
        for (int i = 0; i < arr.length; i++){
            List<Integer> list = idxMap.getOrDefault(arr[i], new ArrayList<>());
            list.add(i);
            idxMap.put(arr[i], list);
        }
        Deque<Integer> queue = new ArrayDeque();
        Set<Integer> walked = new HashSet<>();
        queue.addLast(0);
        while (!queue.isEmpty()){
            int size = queue.size();
            for (int j = 0; j < size; j++){
                int idx = queue.pollFirst();
                walked.add(idx);
                if (idx == n - 1) return step;
                List<Integer> list = idxMap.getOrDefault(arr[idx], new ArrayList<>());
                for (int m: list){
                    if (walked.contains(m)) continue;
                    queue.addLast(m);
                }
                idxMap.remove(arr[idx]);
                if (idx - 1 >= 0 && !walked.contains(idx - 1)){
                    queue.addLast(idx - 1);
                }
                if (idx + 1 < n && !walked.contains(idx + 1)){
                    queue.addLast((idx + 1));
                }
            }
            step++;
        }
        return -1;
    }
}

File: python/1345.跳跃游戏.py
class Solution:
    # 1. O(n) 328ms O(n) 29.2M(59%)
    def minJumps(self, arr: List[int]) -> int:
        hash = defaultdict(list)
        for i, num in enumerate(arr):
            hash[num].append(i)
        walked = set()
        queue = deque([0])
        walked.add(0)
        step = 0
        while queue:
            size = len(queue)
            for iter in range(size):
                node = queue.popleft()
                walked.add(node)
                if node == len(arr)-1: return step
                for n in hash[arr[node]]:
                    if n not in walked:
                        queue.append(n)
                        walked.add(n)
                del hash[arr[node]]
                if 0 <= node - 1 and node - 1 not in walked:
                    queue.append(node - 1)
                    walked.add(node - 1)
                if node + 1 < len(arr) and node + 1 not in walked:
                    queue.append(node + 1)
                    walked.add(node + 1)
            step += 1